relays/bin-substrate/src/cli/init_bridge.rs

```diff
@@ -52,188 +61,42 @@ pub enum InitBridgeName {
-
-				fn encode_init_bridge(
-					init_data: InitializationData<<Source as ChainBase>::Header>,
-				) -> <Target as Chain>::Call {
-					relay_rococo_client::runtime::Call::BridgeGrandpaWococo(
-						relay_rococo_client::runtime::BridgeGrandpaWococoCall::initialize(
-							init_data,
-						),
-					)
-				}
-
-				$generic
-			},
-			InitBridgeName::KusamaToPolkadot => {
-				type Source = relay_kusama_client::Kusama;
-				type Target = relay_polkadot_client::Polkadot;
-				type Engine = GrandpaFinalityEngine<Source>;
-
-				fn encode_init_bridge(
-					init_data: InitializationData<<Source as ChainBase>::Header>,
-				) -> <Target as Chain>::Call {
-					relay_polkadot_client::runtime::Call::BridgeKusamaGrandpa(
-						relay_polkadot_client::runtime::BridgeKusamaGrandpaCall::initialize(
-							init_data,
-						),
-					)
-				}
-
-				$generic
-			},
-			InitBridgeName::PolkadotToKusama => {
-				type Source = relay_polkadot_client::Polkadot;
-				type Target = relay_kusama_client::Kusama;
-				type Engine = GrandpaFinalityEngine<Source>;
-
-				fn encode_init_bridge(
-					init_data: InitializationData<<Source as ChainBase>::Header>,
-				) -> <Target as Chain>::Call {
-					relay_kusama_client::runtime::Call::BridgePolkadotGrandpa(
-						relay_kusama_client::runtime::BridgePolkadotGrandpaCall::initialize(
-							init_data,
-						),
-					)
-				}
-
-				$generic
-			},
-			InitBridgeName::MillauToRialtoParachain => {
-				type Source = relay_millau_client::Millau;
-				type Target = relay_rialto_parachain_client::RialtoParachain;
-				type Engine = GrandpaFinalityEngine<Source>;
-
-				fn encode_init_bridge(
-					init_data: InitializationData<<Source as ChainBase>::Header>,
-				) -> <Target as Chain>::Call {
-					let initialize_call = rialto_parachain_runtime::BridgeGrandpaCall::<
-						rialto_parachain_runtime::Runtime,
-						rialto_parachain_runtime::MillauGrandpaInstance,
-					>::initialize {
-						init_data,
-					};
-					rialto_parachain_runtime::SudoCall::sudo {
-						call: Box::new(initialize_call.into()),
-					}
-					.into()
-				}
-
-				$generic
-			},
-		}
-	};
-}
-
-impl InitBridge {
-	/// Run the command.
-	pub async fn run(self) -> anyhow::Result<()> {
-		select_bridge!(self.bridge, {
-			let source_client = self.source.to_client::<Source>().await?;
-			let target_client = self.target.to_client::<Target>().await?;
-			let target_sign = self.target_sign.to_keypair::<Target>()?;
+trait BridgeInitializer: CliBridgeBase
+where
+	<Self::Target as ChainBase>::AccountId: From<<AccountKeyPairOf<Self::Target> as Pair>::Public>,
+{
+	type Engine: Engine<Self::Source>;
+
+	/// Get the encoded call to init the bridge.
+	fn encode_init_bridge(
+		init_data: <Self::Engine as Engine<Self::Source>>::InitializationData,
+	) -> <Self::Target as Chain>::Call;
+
+	/// Initialize the bridge.
+	fn init_bridge(
+		data: InitBridge,
+	) -> Pin<Box<dyn Future<Output = anyhow::Result<()>> + Send + 'static>> {
```

Comment:
Imo it would be better to use `async_trait` here instead of all this `Pin<Box<...>>` stuff :)

Reply:
(there's also couple of other places where imo it shall be replaced)

Reply:
Indeed, it looks better. Done.